File: server/TodoApp/TodoApp.Application/Services/TaskService.cs
using AutoMapper;
using TodoApp.Domain.Contracts.Exception;
using TodoApp.Domain.Contracts.Task;
using TodoApp.Domain.Interfaces.Repositories;
using TodoApp.Domain.Interfaces.Services;
using DomainTask = TodoApp.Domain.Models.Task;

namespace TodoApp.Application.Services;

public class TaskService : ITaskService
{
   private readonly ITaskRepository _taskRepository;
   private readonly IMapper _mapper;
   private readonly IUnitOfWork _unitOfWork;

   public TaskService(ITaskRepository taskRepository, IMapper mapper, IUnitOfWork unitOfWork)
   {
      _taskRepository = taskRepository;
      _mapper = mapper;
      _unitOfWork = unitOfWork;
   }

   public async Task CreateTask(Guid userId, CreateTaskRequest request, CancellationToken cancellationToken)
   {
      var task = new DomainTask
      {
         Id = Guid.NewGuid(),
         Title = request.Title,
         Note = request.Note,
         IsCompleted = false,
         DueTime = request.DueTime,
         UserId = userId,
      };
      
      await _taskRepository.Add(task);
      
      if (request.CategoryIds != null && request.CategoryIds.Count > 0)
      {
         await _taskRepository.CreateCategoryTasks(request.CategoryIds,task.Id,cancellationToken);
      }

      await _unitOfWork.SaveChangesAsync(cancellationToken);
   }

   public async Task UpdateTask(Guid userId, UpdateTaskRequest request, CancellationToken cancellationToken)
   {
      var task = await _taskRepository.FindById(request.TaskId);
      
      if (task == null)
      {
         throw new ApiException("Task wasn't found",404);
      }

      if (task.UserId != userId)
      {
         throw new ApiException("You can't update this task", 403);
      }
      
      await _taskRepository.UpdateCategoryTasks(request.CategoryIds,task.Id,cancellationToken);
      
      task.Title = request.Title;
      task.Note = request.Note;
      task.DueTime = request.DueTime;
      task.IsCompleted = request.IsCompleted ?? task.IsCompleted;
      
      _taskRepository.Update(task);
      await _unitOfWork.SaveChangesAsync(cancellationToken);
   }

   public async Task DeleteTask(Guid userId, Guid taskId, CancellationToken cancellationToken)
   {
      var task = await _taskRepository.FindById(taskId);

      if (task == null)
         return;

      if (task.UserId != userId)
      {
         throw new ApiException("You can't delete this task", 403);
      }
      
      await _taskRepository.DeleteAsync(taskId);
      await _unitOfWork.SaveChangesAsync(cancellationToken);
   }

   public async Task<SearchTaskResponse> GetUserTasks(Guid userId, SearchTaskOptions searchTaskOptions, CancellationToken cancellationToken)
   {
      var (items, totalCount) = await _taskRepository.GetPagedAndFilteredAsync(userId, searchTaskOptions);
      
      var taskDtos = _mapper.Map<List<TaskDto>>(items);
      
      return new SearchTaskResponse
      {
         Items = taskDtos,
         TotalCount = totalCount,
         Page = searchTaskOptions.Page,
         PageSize = searchTaskOptions.PageSize
      };
   }
}

File: web/to-do-app-web/src/app/core/services/task-service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Service } from '@angular/core';
import { environment } from '../../../environments/environment.development';
import { ApiResponse } from '../types/common/apiResponse';
import { SearchTaskResponse, Task } from '../types/task/task';
import { FilterSettings } from '../types/common/filter-settings';
import { CreateTaskRequest } from '../types/task/createTaskRequest';
import { UpdateTaskRequest } from '../types/task/updateTaskRequest';

@Service()
export class TaskService {
    private client = inject(HttpClient)
    private apiUrl = `${environment.apiUrl}/tasks`

    getUserTasks(searchRequest: FilterSettings){
        let params = new HttpParams({
            fromObject: {
                title: searchRequest.title,
                categoryIds: searchRequest.categoryIds,
                page: searchRequest.page,
                pageSize: searchRequest.pageSize,
            }
        });
        
        if(searchRequest.isCompleted !== null){
            params = params.set('isCompleted', searchRequest.isCompleted);
        }
        
        return this.client.get<ApiResponse<SearchTaskResponse>>(`${this.apiUrl}/me`, { params });
    }

    createTask(createRequest: CreateTaskRequest){
        return this.client.post<ApiResponse<null>>(this.apiUrl, createRequest)
    }

    updateTask(updateRequest: UpdateTaskRequest){
        return this.client.put<ApiResponse<null>>(this.apiUrl, updateRequest)
    }


    deleteTask(taskId: string ){
        return this.client.delete<ApiResponse<null>>(`${this.apiUrl}/${taskId}`)
    }


}
